# plotReads.R

#'
#' \code{plotReads} Return a list of domains that the genes in two systems
#'    overlap along with genes that mapped to the same domains
#'
#' @param coor (data frame) Contains exon coordinates of different genes.
#' @param r (data frame) A data frame containing rows of reads. If a value = 1,
#'     then that row of read is mapped to that column of exon. If a value = 0, then that row
#'     of read is not mapped to that column of exon.
#' @param gene (character) The gene ID of the gene of interest
#' @return (ggplot) A splicing graph of the gene of interest with exon counts of given Nanopore data.
#' @author {Tina Lee} (aut)
#' @examples
#' # Read results generated by countReads function and declare gene of interest.
#' # Run the following code:
#' plotReads(coor, r, gene)
#'
#' @export

plotReads <- function(coor, r, gene, alns) {
  r <- r[r$geneId == gene, ]

  # since this gene only has exon 1 to 8, we can remove column exon9, exon10, exon11, exon12
  # remove columns with only NA
  r <- r[,colSums(is.na(r)) == 0]

  idx <- r[, "index"]

  alns <- alns[idx]

  # get coordinates for correpsonding gene
  speExon <- coor[coor$gene_id == gene, ]

  numAlignments <- length(alns)

  cvg <- readCoverage(speExon, alns)

  p <- ggplot(cvg, aes(x=pos, y=count)) +
    expand_limits(y=c(0, 3*numAlignments)) +
    coord_fixed(ratio = 5) +
    geom_area(color="#00AFBB", fill = "#00AFBB")


  # count and plot arc (missing exons in between)
  arc <- data.frame()

  for (i in 1:(ncol(r) - 3)) {
    for (j in 1:(ncol(r) - 3)) {
      if (abs(i - j) > 1 & (i < j)) {
        # get columns i to j
        s <- i + 3
        e <- j + 3
        tmp <- r[,s:e]
        tmp <- tmp[rowSums(tmp) == 2,]
        first <- noquote(paste0("exon", i))
        sec <- noquote(paste0("exon", j))
        tmp <- tmp[tmp[[first]] == 1 & tmp[[sec]] == 1, ]
        count <- nrow(tmp)
        if (count > 0) {
          arc <- rbind(arc, data.frame("start" = speExon$end[i],
                                       "end" = speExon$start[j],
                                       "count" = count))
        }
      }
    }
  }

  # plot arcs
  s <- speExon$start[1]
  e <- speExon$end[nrow(speExon)]


  for (i in 1:nrow(arc)) {
    start <- arc$start[i]
    end <- arc$end[i]
    ystart <- cvg$count[cvg$pos == start]
    yend <- cvg$count[cvg$pos == end]
    scale <- (ystart + yend)/2 + ((end - start)/12)
    p <- p +
      ggplot2::geom_curve(x = start,
                          xend = end,
                          y = ystart,
                          yend = yend,
                          curvature = -0.75,
                          size = 0.005,
                          color="#00AFBB") +
      ggplot2::geom_text(x = (end - start)/2 + start,
                         y = scale,
                         color = "grey45",
                         label = arc$count[i],
                         size = 2.5)
  }

  # add horizontal lines
  hr <- data.frame()

  # get start and end coordinates of horizontal lines
  for (i in 1:(nrow(speExon) - 1)) {
    first <- noquote(paste0("exon", i))
    sec <- noquote(paste0("exon", i+1))
    tmp <- r[r[[first]] == 1 & r[[sec]] == 1, ]
    count <- nrow(tmp)
    hr <- rbind(hr, data.frame("start" = speExon$end[i],
                               "end" = speExon$start[i+1],
                               "count" = count))
  }

  # plot horizontal lines ans add label
  for (i in 1:nrow(hr)) {
    start <- hr$start[i]
    end <- hr$end[i]
    p <- p +
      ggplot2::geom_segment(x = start,
                            xend = end,
                            y = 0,
                            yend = 0,
                            size = 0.17,
                            color="#00AFBB") +
      ggplot2::geom_text(data = hr,
                         x = start + ((end - start)/2),
                         y = 10,
                         color = "grey45",
                         label = hr$count[i],
                         size = 2.5)
  }

  return(p)
}



# [END]
